pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -95,31 +106,95 @@ type DNSProxy struct {
 	// lookupTargetDNSServer extracts the originally intended target of a DNS
 	// query. It is always set to lookupTargetDNSServer in
 	// helpers.go but is modified during testing.
-	lookupTargetDNSServer func(w dns.ResponseWriter) (server string, err error)
+	lookupTargetDNSServer func(w dns.ResponseWriter) (serverIP net.IP, serverPort uint16, addrStr string, err error)
 
 	// this mutex protects variables below this point
 	lock.Mutex
 
-	// allowed tracks all allowed matchNames. These are regexps, even simple
-	// matchNames with no regexp wildcards are compiled. We ensure a unique value
-	// per source policy of a matchName because the RegexpMap handles the
-	// reference counting of unique values but will de-dupe repeats.
+	// allowed tracks all allowed L7 DNS rules by endpointID, destination port,
+	// and L3 Selector. All must match for a query to be allowed.
+	//
+	// matchNames with no regexp wildcards are still compiled, internally.
 	// Note: Simple DNS names, e.g. bar.foo.com, will treat the "." as a literal.
-	// We convert these "." in regexps that only contain dots (and alphanumeric
-	// characters) into a regexp "." literal. This is because the more common use
-	// case will be these simple literals.
-	// To insert a wildcard ".", use .{1} to indicate a single wildcard character.
-	allowed *regexpmap.RegexpMap
+	allowed perEPAllow
 
 	// rejectReply is the OPCode send from the DNS-proxy to the endpoint if the
 	// DNS request is invalid
 	rejectReply int
 }
 
+// perEPAllow maps EndpointIDs to ports + selectors + rules
+type perEPAllow map[uint64]portToSelectorAllow
```

Comment:
Maybe consider tracking ports, selectors, and rules by a SecurityID instead of by EndpointID? Seems we could share the state between endpoints that have the same SecurityID here. There would be some additional complexity though due to policies being created per EP. `policy.L4Filter.L7RulesPerEp` might be a proper place to keep this information. Note that there is an another (wip) PR (https://github.com/cilium/cilium/pull/9486) that generalizes `L7RulesPerEp` a little, allowing for computed state being stored there, not just the `api.L7Rules`.